File: SerpentFrameworkBot/src/main/cpp/serpentframework/Waypoint.cpp
#include "serpentframework/Waypoint.h"
#include <math.h>
#define M_PI 3.14159265358979323846

serpentframework::Waypoint::Waypoint(char i, WaypointType action, double x, double z) {
    id = i;
    actionType = action;
    xPosition = x;
    zPosition = z;
}

double serpentframework::Waypoint::distance(Waypoint a, Waypoint b) {
    return sqrt(pow(a.xPosition - b.xPosition, 2) + pow(a.zPosition - b.zPosition, 2));
}

serpentframework::Quaternion2D serpentframework::Waypoint::quaternionAngle(Waypoint a, Waypoint b) {
    return serpentframework::Quaternion2D::fromAxis(b.xPosition - a.xPosition, b.zPosition - a.zPosition);
}

double serpentframework::Waypoint::eulerAngle(Waypoint a, Waypoint b) {
   return (-2 * signbit(b.xPosition - a.xPosition) + 1) * asin(b.zPosition - a.zPosition) * (180 / M_PI);
}

File: SerpentFrameworkBotJ/src/main/java/org/swampscottcurrents/serpentframework/Waypoint.java
package org.swampscottcurrents.serpentframework;

public class Waypoint {
    
    private static byte pointCount;

    public final byte id;
    public double xPosition;
    public double yPosition;
    public WaypointType actionType;

    public Waypoint() {
        id = pointCount;
        actionType = WaypointType.NAVIGATION;
        pointCount++;
    }

    public Waypoint(WaypointType action, double x, double y) {
        id = pointCount;
        xPosition = x;
        yPosition = y;
        actionType = action;
        pointCount++;
    }

    /** Gets the absolute angle between two Waypoints, where Waypoint a is the origin. */
    public static Quaternion2D getAngle(Waypoint a, Waypoint b) {
        return Quaternion2D.fromAxis(b.xPosition - a.xPosition, b.yPosition - a.yPosition);
    }

    /** Gets the distance between two Waypoints. */
    public static double getDistance(Waypoint a, Waypoint b) {
        return Math.sqrt(square(a.xPosition - b.xPosition) + square(a.yPosition - b.yPosition));
    }

    private static double square(double a) {
        return a * a;
    }
}